exporter/tanzuobservabilityexporter/transformer.go

```diff
@@ -0,0 +1,228 @@
+		tags[k] = v
+	}
+
+	if len(orig.TraceState()) > 0 {
+		tags[tracetranslator.TagW3CTraceState] = string(orig.TraceState())
+	}
+
+	return Span{
+		Name:           orig.Name(),
+		TraceID:        traceID,
+		SpanID:         spanID,
+		ParentSpanID:   parentSpanID,
+		Tags:           tags,
+		StartMillis:    startMillis,
+		DurationMillis: durationMillis,
+		SpanLogs:       eventsToLogs(orig.Events()),
+	}, nil
+}
+
+func spanKind(span pdata.Span) string {
+	switch span.Kind() {
+	case pdata.SpanKindCLIENT:
+		return "client"
+	case pdata.SpanKindSERVER:
+		return "server"
+	case pdata.SpanKindPRODUCER:
+		return "producer"
+	case pdata.SpanKindCONSUMER:
+		return "consumer"
+	case pdata.SpanKindINTERNAL:
+		return "internal"
+	case pdata.SpanKindUNSPECIFIED:
+		return "unspecified"
+	default:
+		return "unknown"
+	}
+}
+
+func (t *traceTransformer) setRequiredTags(tags map[string]string) {
+	if _, ok := tags[labelService]; !ok {
+		if _, svcNameOk := tags[conventions.AttributeServiceName]; svcNameOk {
+			tags[labelService] = tags[conventions.AttributeServiceName]
+			delete(tags, conventions.AttributeServiceName)
+		} else {
+			tags[labelService] = defaultServiceName
+		}
+	}
+	if _, ok := tags[labelApplication]; !ok {
+		tags[labelApplication] = defaultApplicationName
+	}
+}
+
+func eventsToLogs(events pdata.SpanEventSlice) []senders.SpanLog {
+	var result []senders.SpanLog
+	for i := 0; i < events.Len(); i++ {
+		e := events.At(i)
+		fields := attributesToTags(e.Attributes())
+		fields[labelEventName] = e.Name()
+		result = append(result, senders.SpanLog{
+			Timestamp: int64(e.Timestamp()) / time.Microsecond.Nanoseconds(), // Timestamp is in microseconds
+			Fields:    fields,
+		})
+	}
+
+	return result
+}
+
+func calculateTimes(span pdata.Span) (int64, int64) {
+	startMillis := int64(span.StartTimestamp()) / time.Millisecond.Nanoseconds()
+	endMillis := int64(span.EndTimestamp()) / time.Millisecond.Nanoseconds()
+	durationMillis := endMillis - startMillis
+	// it's possible end time is unset, so default to 0 rather than using a negative number
+	if span.EndTimestamp() == 0 {
+		durationMillis = 0
+	}
+	return startMillis, durationMillis
+}
+
+func attributesToTags(attributes ...pdata.AttributeMap) map[string]string {
+	tags := map[string]string{}
+
+	extractTag := func(k string, v pdata.AttributeValue) bool {
+		tags[k] = tracetranslator.AttributeValueToString(v, false)
+		return true
+	}
+
+	// Since AttributeMaps are processed in the order received, later values overwrite earlier ones
+	for _, att := range attributes {
+		att.Range(extractTag)
+	}
+
+	return tags
+}
+
+func errorTagsFromStatus(status pdata.SpanStatus) map[string]string {
+	tags := map[string]string{
+		labelStatusCode: fmt.Sprintf("%d", status.Code()),
+	}
+	if status.Code() != pdata.StatusCodeError {
+		return tags
+	}
+
+	tags[labelError] = "true"
+	if status.Message() != "" {
+		msg := status.Message()
+		maxLength := 255 - len(labelStatusMessage+"=")
```

Comment:
I think you can make this const
```suggestion
		const maxLength = 255 - len(labelStatusMessage+"=")
```